### PR TITLE
Keyboard text selection and wordwrap

### LINKED ISSUE #79291: MacOS move selection by line inconsistency
The cmd + left/right arrow shortcut is able to jump between lines in Flutter, but when using TextEdit on Mac, the cursor stops at the beginning/end of the line and isn't able to change lines.

## Steps to reproduce

1. Run an app with a multi-line text field, such as the one below, on a Mac.
2. Type a few lines of text and put the cursor at the end.
3. Hold the cmd key and press the left arrow twice.

Expected results: When doing this in TextEdit, the second keypress has no effect.  The first one moves the cursor to the start of the last line, and the second one does nothing.

Actual results: The cursor moves to the start of the second to last line.

<details>

<summary>Code</summary>

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Scaffold(
        appBar: AppBar(
          title: Text(widget.title),
        ),
        body: Center(
          child: TextField(
            maxLines: 5,
          ),
        ),
      ),
    );
  }
}
```

</details>

### PR DESCRIPTION
**Pull request by justinmc**
Expanding by line (e.g. cmd-shift-left/right) shouldn't jump between lines when hit multiple times.  According to my testing on native, it should go to the start/end of the line, and subsequent presses should do nothing.  This PR matches that behavior and fixes some other sketchiness around the edges of lines, thanks to the engine bugfix referenced below.

After the PR it seems to work properly:

wordwrap

Fixes https://github.com/flutter/flutter/issues/79291
Depends on https://github.com/flutter/engine/pull/28008

### CODE REVIEW

File: packages/flutter/lib/src/rendering/editable.dart [hunk 1]
```diff
@@ -1939,6 +1949,15 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin,
       return moveSelectionRightByLine(cause);
     }
 
+    // If greatest edge is already at the end of a line, don't do anything.
+    final int greatestOffset = math.max(selection!.baseOffset, selection!.extentOffset);
+    final TextSelection currentLine = _getLineAtOffset(TextPosition(
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/rendering/editable.dart:1954]: Could the problem be we're initializing a new `TextPosition` with the default text affinity being downstream? If text affinity is upstream then `_getLineAtOffset` won't go to the next line right?

**Comment by goderbauer** [comment]
(PR triage): @justinmc Do you still have plans for this PR?

**Comment by justinmc** [comment]
@goderbauer Yes I do!  I was waiting for https://github.com/flutter/flutter/pull/87133 to be merged.  Coming back to this this week now.

**Comment by justinmc** [comment]
This is blocked on https://github.com/flutter/engine/pull/28008.  There is a bug in the engine that should be fixed before this is fixed here, otherwise I'd be hacking around a bug.

**Comment by justinmc** [comment]
Squashed commits in order to fix svg test.

**Comment by justinmc** [comment]
@LongCatIsLooong Ready for re-review.  This bumps into the problem of not having multiple TextAffinities for a non-collapsed TextSelection, which you mentioned [here](https://github.com/flutter/flutter/pull/87973/files#r685632915) and I wrote up in https://github.com/flutter/flutter/issues/88135.  I think this PR works in all practical situations though, which I've covered in my test, even though we're not able to distinguish certain non-collapsed selections.

File: packages/flutter/lib/src/rendering/editable.dart [hunk 2]
```diff
@@ -1747,6 +1747,16 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin,
       return moveSelectionLeftByLine(cause);
     }
 
+    // If the lowest edge of the selection is at the start of a line, don't do
+    // anything.
+    final int lowestOffset = math.min(selection!.baseOffset, selection!.extentOffset);
```
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/rendering/editable.dart:1752]: these lines seem to be doing the same as line 1760-1762?

**Review by LongCatIsLooong — APPROVED** [review]

**Pull request merged**